_drafts/makefiles.md

```diff
@@ -0,0 +1,329 @@
+what the target, prerequisites, and the build rules would be.
+
+<div class="blank-page"></div>
+
+Done? Let's go over it. The target is the thing we want to end up making. In our
+case, let's call it `submit.tar.gz`. So our rule so far looks like this:
+
+{% highlight make %}
+submit.tar.gz:
+{% endhighlight %}
+
+What are its prerequisites? Or, if you like to think about it this way, when
+would we need to regenerate our submit file? We need to regenerate the submit
+file when any its contents change. This includes our header and source files,
+and also our `Makefile` and test cases.
+
+We say that `submit.tar.gz` *depends on* our headers, sources, and `Makefile`.
+
+<aside class="aside-info"><p>
+
+The terms "prerequisite" and "dependency" are interchangeable. Just to remind
+you, these are the things that we need to be present and up-to-date before we
+can start making the target file.
+
+</p></aside>
+
+Let's add them to our rule:
+
+{% highlight make %}
+submit.tar.gz: main.cpp main.h test-1.txt Makefile # And anything else...
+{% endhighlight %}
+
+So with this rule, we're telling `make` that it's clear to try building the
+`submit.tar.gz` file as long as `main.cpp`, `main.h`, etc. 1) exist on the file
+system, and 2) are up-to-date. For now, we'll ignore the idea of `main.cpp` and
+friends being up to date or not. Just pretend that they are.
+
+The last thing we need to do is to tell `make` how to actually create
+`submit.tar.gz` given all of these files. We'll invoke the `tar` command to do
+that.
+
+{% highlight make %}
+submit.tar.gz: main.cpp main.h test-1.txt Makefile # And anything else...
+	tar -czf submit.tar.gz main.cpp main.h test-1.txt Makefile
+{% endhighlight %}
+
+<aside class="aside-warning"><p>
+
+We're promising `make` that after it's done running all of our commands,
+`submit.tar.gz` will be created and up-to-date. `make` won't verify that you
+actually did, so if you didn't actually build `submit.tar.gz`, you would
+probably have issues.
+
+</p></aside>
+
+Try putting this content in your `Makefile`, in the same directory as whatever
+source files you might have. Then, in that same directory, run `make
+submit.tar.gz` from the command line. It should run the `tar` command the first
+time, and skip it each time after that. Try editing one of your files and run
+`make submit.tar.gz` again --- it should regenerate your `.tar.gz` file.
+
+# Variables in Makefiles
+
+Writing out all of our submission files is difficult to maintain if we want to
+update them. So we can write this instead:
+
+{% highlight make %}
+SUBMIT_FILES = main.cpp main.h test-1.txt Makefile
+
+submit.tar.gz: $(SUBMIT_FILES)
+	tar -czf $(SUBMIT_FILES)
+{% endhighlight %}
+
+But this still means that we have to update this file whenever we add files. We
+can use `make`'s convenient `wildcard` function to address this:
+
+{% highlight make %}
+SUBMIT_FILES = $(wildcard *.cpp *.h test-*.txt) Makefile
```

Comment:
Github's markdown renderer doesn't handle Jekyll directives — it ends up working fine since it's in a `highlight` block:

![screen shot 2015-08-29 at 5 35 43 pm](https://cloud.githubusercontent.com/assets/454057/9564408/6f3ea5c6-4e74-11e5-8d2a-8e29160bf09a.png)